Sources/SwiftSyntax/SyntaxData.swift

```diff
@@ -79,35 +81,66 @@ struct AbsoluteSyntaxInfo {
   }
 }
 
+/// Represents a unique value for a node within its own tree.
+struct SyntaxIndexInTree: Hashable {
+  let indexInTree: UInt32
+
+  static var zero: SyntaxIndexInTree = SyntaxIndexInTree(indexInTree: 0)
+
+  func advancedBySibling(_ raw: RawSyntax?) -> SyntaxIndexInTree {
+    let newIndexInTree = self.indexInTree + UInt32(truncatingIfNeeded: raw?.totalNodes ?? 0)
+    return .init(indexInTree: newIndexInTree)
+  }
+
+  func reversedBySibling(_ raw: RawSyntax?) -> SyntaxIndexInTree {
+    let newIndexInTree = self.indexInTree - UInt32(truncatingIfNeeded: raw?.totalNodes ?? 0)
+    return .init(indexInTree: newIndexInTree)
+  }
+
+  func advancedToFirstChild() -> SyntaxIndexInTree {
+    let newIndexInTree = self.indexInTree + 1
+    return .init(indexInTree: newIndexInTree)
+  }
+
+  func advancedToEndOfChildren(_ raw: RawSyntax) -> SyntaxIndexInTree {
+    let newIndexInTree = self.indexInTree + UInt32(truncatingIfNeeded: raw.totalNodes)
+    return .init(indexInTree: newIndexInTree)
+  }
+
+  private init(indexInTree: UInt32) {
+    self.indexInTree = indexInTree
+  }
+}
+
 /// Provides a stable and unique identity for `Syntax` nodes.
 public struct SyntaxIdentifier: Hashable {
```

Comment:
Not related to the overall PR but it occurred to me that this could use the `Identifiable` protocol.

Reply:
I thought the same and then forgot about it. Will do so in a follow-up PR.